source/grid/manifold_lib.cc

```diff
@@ -549,6 +551,284 @@ TorusManifold<dim>::push_forward_gradient(const Point<3> &chart_point) const
+      new_point += my_weight * cell->face(face)->get_manifold().get_new_point(points, weights);
+    }
+  return new_point;
+}
+
+
+
+template <int dim, int spacedim>
+DerivativeForm<1,dim,spacedim>
+TransfiniteInterpolationManifold<dim,spacedim>
+::push_forward_gradient(const typename Triangulation<dim,spacedim>::cell_iterator &cell,
+                        const Point<dim> &chart_point) const
+{
+  // compute the derivative with the help of finite differences
+  const double eps = 1e-8;
+  Point<spacedim> point = push_forward(cell, chart_point);
+  DerivativeForm<1,dim,spacedim> grad;
+  for (unsigned int d=0; d<dim; ++d)
+    {
+      Point<dim> modified = chart_point;
+
+      // avoid checking outside of the unit interval
+      if (chart_point[d] > 0.5)
+        modified[d] -= eps;
+      else
+        modified[d] += eps;
+      Tensor<1,spacedim> difference = push_forward(cell, modified) - point;
+      for (unsigned int e=0; e<spacedim; ++e)
+        grad[e][d] = difference[e] / eps * (chart_point[d] > 0.5 ? -1. : 1.);
+    }
+  return grad;
+}
+
+
+
+template <int dim, int spacedim>
+Point<dim>
+TransfiniteInterpolationManifold<dim,spacedim>
+::pull_back(const typename Triangulation<dim,spacedim>::cell_iterator &cell,
+            const Point<spacedim> &point) const
+{
+  Point<dim> chart_point;
+  for (unsigned int d=0; d<dim; ++d)
+    chart_point[d] = 0.5;
+
+  // run Newton iteration
+  for (unsigned int i=0; i<40; ++i)
+    {
+      Tensor<1,spacedim> residual = point - push_forward(cell, chart_point);
+      const double residual_norm = residual.norm();
+      if (residual_norm < 1e-13)
+        break;
+
+      DerivativeForm<1,dim,spacedim> inv_grad =
+        push_forward_gradient(cell, chart_point).covariant_form();
+      Tensor<1,dim> update;
+      for (unsigned int d=0; d<spacedim; ++d)
+        for (unsigned int e=0; e<dim; ++e)
+          update[e] += inv_grad[d][e] * residual[d];
+
+      // Reduce step length in case the new point is outside the unit cell
+      double alpha = 1.;
+      while (GeometryInfo<dim>::is_inside_unit_cell(chart_point + alpha*update,
+                                                    1e-7) == false
+             &&
+             alpha > 1e-5)
+        alpha *= 0.5;
+      chart_point += alpha * update;
+    }
+  return chart_point;
+}
+
+
+
+template <int dim, int spacedim>
+Point<spacedim>
+TransfiniteInterpolationManifold<dim, spacedim>
+::get_new_point (const std::vector<Point<spacedim> > &surrounding_points,
+                 const std::vector<double>           &weights) const
+{
+  const typename Triangulation<dim,spacedim>::cell_iterator cell =
+    this->get_cell_around_points(surrounding_points);
+
+  std::vector<Point<dim> > chart_points(surrounding_points.size());
+
+  for (unsigned int i=0; i<surrounding_points.size(); ++i)
+    chart_points[i] = pull_back(cell, surrounding_points[i]);
+
+  FlatManifold<dim> sub_manifold;
```

Comment:
Yes, maybe so.

My worries when I see this is always "what happens if `FlatManifold` suddenly acquires state and becomes more expensive?". That's unlikely to happen for `FlatManifold`, but we've come up with lots of things that had simple implementations in the beginning and then later on become significantly more complicated. We can guard against that by making sure we don't create too many *potentially* expensive objects here and there.

As I said, it's not likely going to be a problem here. Do as you see fit.